src/engine/net/Connection.cpp

```diff
@@ -0,0 +1,266 @@
+#include "Connection.hpp"
+#include <iostream>
+
+using namespace admirals::net;
+
+Connection::Connection(Owner parent, asio::io_context &io_context,
+                       asio::ip::tcp::socket socket,
+                       MessageQueue<OwnedMessage> &incoming_messages,
+                       Server *server, ValidationFunction validation_function)
+    : m_ioContext(io_context), m_socket(std::move(socket)),
+      m_incomingMessages(incoming_messages) {
+
+    m_ownerType = parent;
+    m_server = server;
+    m_validationFunction = validation_function;
+
+    if (m_ownerType == Owner::SERVER) {
+        // Uses the current time as the validation number
+        m_validationOut = uint64_t(
+            std::chrono::system_clock::now().time_since_epoch().count());
+    }
+}
+
+Connection::~Connection() {}
+
+void Connection::ConnectToClient(admirals::net::Server *server, uint32_t uid) {
+    if (m_socket.is_open()) {
+        m_id = uid;
+        // Send the validation number to the client and wait for the client to
+        // send it back hashed
+        WriteValidation();
+        ReadValidation();
+    }
+}
+
+void Connection::ConnectToServer(
+    const asio::ip::tcp::resolver::results_type &endpoints) {
+    asio::async_connect(
+        m_socket, endpoints,
+        [this](std::error_code ec, asio::ip::tcp::endpoint endpoint) {
+            if (!ec) {
+                // Read the validation number from the server and send it back
+                // hashed
+                ReadValidation();
+            } else {
+                std::cout << "Connect failed: " << ec.message() << std::endl;
+            }
+        });
```

Comment:
I think we should extract this lambda expression. This one is not that badly formatted, but i think we should stay away from lambda expressions with multiline bodies.

Reply:
Changed in 713165b